### PR TITLE
Introduce VirtualMachine status field and CLI column

### PR DESCRIPTION
**Pull request by zcahana**
**What this PR does / why we need it**:

This PR adds a new `status.printableStatus` field to the VirtualMachine CRD, as well as a new `STATUS` column to the output of `kubectl get vm`.

The purpose of this field is to provide a high-level representation of the status of the virtual machine. It's designated to be consumed by human users, and not automated processes (i.e., no state machine semantics, values may be added or dropped in the future, ...).

Here's an example output:
```bash
$ kubectl get vms
NAME     AGE   STATUS    VOLUME
testvm   4s    Stopped
```

**Release note:**
```release-note
Introduced a "status.printableStatus" field in the VirtualMachine CRD. This field is now displayed in the tabular output of "kubectl get vm".
```

### CODE REVIEW

**Comment by zcahana** [comment]
/hold

This PR still lacks automated testing.
While I work on adding that, it could be useful to have a pair of eyes start reviewing this.

**Comment by maiqueb** [comment]
Would you mind pasting an example output to the PR's description ?

**Comment by zcahana** [comment]
> Would you mind pasting an example output to the PR's description ?

@maiqueb Added, see PR description.

**Comment by ezrasilvera** [comment]
@zcahana As this is printable status aimed for the user, should it be documented somewhere? i.e., what are the possible values and what each value exactly means?  (cc: @rmohr - is there some user-facing guide that can/should include this ? )

File: pkg/virt-controller/watch/vm_test.go
Comment (ezrasilvera) [inline pkg/virt-controller/watch/vm_test.go:1192]: It seems that many of the different tests below are very similar. Maybe you can create a more general table function that will allow you to put most of these tests into a single table ? 
For example "should set a Stopping status when VMI has a deletion timestamp set"   and ""should set a Terminating status when VM has a deletion timestamp set"  seems almost identical ...

**Comment by zcahana** [comment]
There's some documentation at the [type and values definitions](https://github.com/kubevirt/kubevirt/pull/5528/commits/8f2f287ff366771444a70e4af5c79032b0c348cc#diff-5469726ef6ef072d9258db4593eb8811363099d9181a4e5dbd119af7b2b1414eR1111), and at the [field declaration](https://github.com/kubevirt/kubevirt/pull/5528/commits/8f2f287ff366771444a70e4af5c79032b0c348cc#diff-5469726ef6ef072d9258db4593eb8811363099d9181a4e5dbd119af7b2b1414eR1157). As such, it should be visible (only the field documentation) via `kubectl explain virtualmachine.status`, and I also expect it to be picked up into the API reference page (need to validate that it happens automatically).

Do you think this merits more than that, i.e., some section somewhere on the [user guide](https://kubevirt.io/user-guide/)? In the meantime I've scanned the user-guide and some other docs, and didn't find any CLI examples that need to be adapted with the new CLI output.

Also, we should strives to keep the values self-explanatory a much as possible (e.g., you don't need a document specifying what's ErrImagePull is for a pod, right). One not-so-self-explanatory status is Provisioning vs. Starting, i.e. it's not entirely clear what's the distinction.

File: tests/kubectl_test.go
Comment (ezrasilvera) [inline tests/kubectl_test.go:139]: Was that a duplicated test? Or was it removed for something related to this PR?

File: pkg/virt-controller/watch/vm_test.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm_test.go:1192]: There are subtle differences between these tests so most of them ended up in `It()`s of their own. I tried to gather to `table.DescribeTable` where it made sense. Nevertheless I'll go through it again and see if I can gather up a few more.

File: tests/kubectl_test.go
Comment (zcahana) [reply tests/kubectl_test.go:139]: Yes, it was byte-identical to the test case defined just before it.

Comment (ezrasilvera) [reply tests/kubectl_test.go:139]: Thanks. Strange that Sonar cloud didn't pick it :(

File: pkg/virt-controller/watch/vm.go
Comment (ezrasilvera) [inline pkg/virt-controller/watch/vm.go:1409]: I put the Q here - but it's a general one - Are you sure the status conditions are all mutual exclusive ?

Comment (zcahana) [reply pkg/virt-controller/watch/vm.go:1409]: No, see the code comment just above that:
```go
// For each status, there's a separate function that evaluates
// whether the status is "true" for the given VM.
//
// Note that these statuses aren't mutually exclusive,
// and several of them can be "true" at the same time
// (e.g., Running && Migrating, or Paused && Terminating).
//
// The actual precedence of these statuses are determined by the order
// of evaluation - first match wins.
```

File: tests/kubectl_test.go
Comment (zcahana) [reply tests/kubectl_test.go:139]: Indeed SonarCloud is configured to exclude `*_test.go` files. As if they don't need to be maintained :-)

File: pkg/virt-controller/watch/vm.go
Comment (ezrasilvera) [reply pkg/virt-controller/watch/vm.go:1409]: This seems problematic isn't it ?  So in theory someone can run the `get` twice on a VM while nothing changed but still get two different status outcomes ?

Comment (zcahana) [reply pkg/virt-controller/watch/vm.go:1409]: The above comment relates to these "status funcs" being not mutually exclusive. It means that two or more of these functions can (potentially) evaluate to true at the same time, and hence their order of evaluation is important.

File: pkg/virt-controller/watch/vm.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vm.go:772]: This is quite the optimization.

File: pkg/virt-controller/watch/vm_test.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vm_test.go:696]: maybe we should be setting the vm.Status.PrintableStatus to the expected value here to avoid the update always occurring. This changes the meaning of the test a bit by always having the UpdateStatus hit.

I think this goes for all the existing tests where `vmInterface.EXPECT().UpdateStatus(gomock.Any()).Times(1).Return(vm, nil)` has been added to satisfy the PrintableStatus being set during execute()

Not having an UpdateStatus() client call invoked is part of these test cases. We want to make sure that UpdateStatus is only called when something expected changes. By having it always be expected due to the PrintableStatus change, then we could accidently leak in other unexpected changes as well that would invoke UpdateStatus.

does that make sense?

**Review by davidvossel — COMMENTED** [review]
excellent job! This is pretty much spot on. I just had one comment about the unit testing.

File: pkg/virt-controller/watch/vm_test.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm_test.go:696]: Yes, that makes perfect sense. I'll modify accordingly.

File: pkg/virt-controller/watch/vm.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm.go:772]: In the sense that it's not needed? It basically allows to detect a status change a bit earlier.

File: pkg/virt-controller/watch/vm_test.go
Comment (maiqueb) [inline pkg/virt-controller/watch/vm_test.go:1216]: I think you shouldn't use a boolean argument, there's hardly ever a motivation for it.

In this case, you can achieve the exact same passing directly a `*Time` struct.

File: pkg/virt-controller/watch/vm_test.go
Comment (maiqueb) [inline pkg/virt-controller/watch/vm_test.go:1263]: I can't fail to notice that there's an input parameter for `Running` but you're passing it `true` in all test entries; why bother ? Couldn't you just omit the parameter and assume it'll always be true ?

**Review by maiqueb — COMMENTED** [review]
Looks nice, thorough coverage on the unit tests.

My comments are pretty much just nits.

File: pkg/virt-controller/watch/vm_test.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm_test.go:1263]: Well, I failed to notice that :-)
Will remove.

**Comment by ezrasilvera** [comment]
I think that the fact we print partial state information because some states can co-exists might be problematic. For example I think it's important for the user to know if we are in "terminating" and "stopped"  or "terminating" and "running" , with the above printout he will get only "terminating" . This is true for other combinations as well (migrating + stopped/running, etc..) . Note, that some of theses sates have big impact on the user (e.g., terminating + stopped --> guaranteed that the app is down, as oppose to terminating + running )
I'm not a user experience expert so I'm not sure what is the best solution here - the only easy thing I can think of is print all the available states as a comma-separated list , or (less preferable to my opinion) have dedicated column for Migrating  (true/false) and Terminating (true/false) 
@davidvossel @rmohr - what do you think?  You probably know better what is the user-experience we want in KubeVirt

File: pkg/virt-controller/watch/vm_test.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm_test.go:1216]: Good point. Will change to `*Time`.

File: pkg/virt-controller/watch/vm_test.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm_test.go:1194]: Having another look, I think that merging more of these tests to parameterized tables would yield test code with too many conditionals and/or obscure testcase parameters. I prefer to leave as is, despite the inevitable duplication.

File: pkg/virt-controller/watch/vm_test.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm_test.go]: btw, fixing this made me realize that the current implementation doesn't detect when a VM with pending dataVolumeTemplates is being started (i.e., it stays in Stopped status up until the DVs are ready and the VMI gets created, and only then moves to Provisioning). I'll work on fixing that too.

**Comment by zcahana** [comment]
Added a section to the user-guide: kubevirt/user-guide#439

**Comment by zcahana** [comment]
/hold

Realized that Provisioning status is still incorrectly detected when there are in-progress DVs.
Working on a fix for that.

File: pkg/virt-controller/watch/vm.go
Comment (davidvossel) [reply pkg/virt-controller/watch/vm.go:789]: yup, i think it's fine. i was merely surprised to see it. Using expectaions like this was something I hadn't considered before

**Comment by davidvossel** [comment]
>I think that the fact we print partial state information because some states can co-exists might be problematic. For example I think it's important for the user to know if we are in "terminating" and "stopped" or "terminating" and "running" , with the above printout he will get only "terminating" . This is true for other combinations as well (migrating + stopped/running, etc..) . Note, that some of theses sates have big impact on the user (e.g., terminating + stopped --> guaranteed that the app is down, as oppose to terminating + running )

> @davidvossel - what do you think? You probably know better what is the user-experience we want in KubeVirt

By reducing the human readable state down to a single value, I think this is just something we have to deal with. There's only so much fidelity that can be expressed in one value effectively. 

When I look at these overlapping values, it's usually pretty obvious which one is the most pertinent one to a user though. 

For example, if we're both migrating and terminating, then terminating is the one I'd want to see.  If we're terminating and the compute container has technically stopped, I'd still want to see terminating because we're waiting for the entire environment to get torn down. If the VMI is paused and migrating, i'd rather see migrating and then for it to return to paused after the migration completes. 

in short, i think we're on the right track with what we have right now. States will overlap, we just have to make some decisions about what is most pertinent to the user.

**Comment by zcahana** [comment]
> Realized that Provisioning status is still incorrectly detected when there are in-progress DVs.
Working on a fix for that.

I've reworked how we detect the Provisioning status. Previously, we were setting it only when the VMI existed and either was still "unprocessed" or had its Provisioning condition set; This was only working for WaitForFirstConsumer volumes. Now, we set Provisioning whenever there are DataVolumes being prepared, regardless if the VMI exists or not (even when it has `running: false`). I believe this better reflects what's going on.

Also reworked some of the unit tests, added cases for the above, and removed some redundant ones.

**Comment by zcahana** [comment]
Note that last commit, ae39760, is a cherry-pick from #5595 which isn't merged just yet (otherwise we have a unit test failing). Once it's merged I can rebase and remove it from here.

**Comment by zcahana** [comment]
@davidvossel @ezrasilvera @maiqueb I believe this is quite ready now and all review comments have been addressed, please let me know if there's anything else needed here, thanks!

File: pkg/virt-controller/watch/vm.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vm.go:581]: There's no rule the a DataVolume has to be defined in the DataVolumeTemplate section on a VM or that every DataVolume used by a VM is controlled by the corresponding VM

I think we can simplify this check and make it more accurate. Instead of looping through the DataVolumeTemplate list, loop through the vm.spec.template.spec.volumes list. For every DataVolume found there, find it in the cache, and simply verify that `dv.Status.Phase == cdiv1.Succeeded`

File: pkg/virt-controller/watch/vm.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vm.go:553]: i'd only run this check if the VM's VMI is not phase == running. Due to hotplug we can provision storage while the VMI is running, and i think that would get confusing to see a VM that's currently active (running) being reported as provisioning. 



Also, It's technically possible for someone to delete a DataVolume while the VMI is running, and I don't know what that does to the DataVolume's phase. The underlying PVC won't be removed until the VMI's pod exits though, so it wouldn't strip it out from under the VMI.

**Review by davidvossel — COMMENTED** [review]
i found a couple of minor things related to the datavolume checks.

File: pkg/virt-controller/watch/vm.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm.go:553]: > i'd only run this check if the VM's VMI is not phase == running. Due to hotplug we can provision storage while the VMI is running, and i think that would get confusing to see a VM that's currently active (running) being reported as provisioning.

Done. 

> Also, It's technically possible for someone to delete a DataVolume while the VMI is running, and I don't know what that does to the DataVolume's phase. The underlying PVC won't be removed until the VMI's pod exits though, so it wouldn't strip it out from under the VMI.

Just tried that, and it does nothing to the DataVolume's phase. Consequently, it does nothing to the VM's status too - I believe this is the right behavior. If the VMI is stopped and the DataVolume deletes successfully, we'd be back to Provisioning status.

File: pkg/virt-controller/watch/vm.go
Comment (zcahana) [reply pkg/virt-controller/watch/vm.go:581]: That's a good point. I've reworked the logic to loop through `vm.spec.template.spec.volumes`.

It is, however, might be misleading, if we display a "Provisioning" status for a VM with a referenced DataVolume that doesn't exist, as there's nothing in the system that actively tries to provision that DataVolume. In that case it may be better to display a different status that highlights the fact that the system is waiting for some action (DataVolume creation) to be taken.

Comment (zcahana) [reply pkg/virt-controller/watch/vm.go:581]: And another though - we may also need to examine volumes of type persistentVolumeClaim, and verify whether they are bound (ready), in-progress (--> Provisioning status), or just don't match anything (which is quite similar to the above case with DVs).

**Review by davidvossel — APPROVED** [review]
/lgtm

File: pkg/virt-controller/watch/vm.go
Comment (davidvossel) [reply pkg/virt-controller/watch/vm.go]: > If you decide to introspect PVCs, the basic check is confirming that the PVC is bound. From there if the PVC is owned by a DataVolume, you could cross reference the DataVolume to determine if the PVC is populated as well.

just to clarify though. I don't think this is really necessary. I've (lgtm) the pr, you can follow up wiht this change if you want. It doesn't matter much.

Comment (zcahana) [reply pkg/virt-controller/watch/vm.go]: I've indeed added the PVC-volumes check. I don't check the DV owner of the PVC, since in the first place I'm only directly reaching to PVCs which are referenced from volumes of type PersistentVolumeClaim (and DVs are reached directly for volume of type DataVolume...).

**Comment by zcahana** [comment]
> /hold
>
> Realized that Provisioning status is still incorrectly detected when there are in-progress DVs.
Working on a fix for that. 

/hold cancel

**Comment by zcahana** [comment]
@davidvossel thanks for reviewing this! Please let me know if there anything else needed for /approval.

**Comment by zcahana** [comment]
> pull-kubevirt-e2e-k8s-1.19-sig-compute Expected — Waiting for status to be reported

This test is stuck in pending state for a couple of days now...
/retest

**Pull request merged**
